src/hardware/disney.cpp

```diff
@@ -19,456 +20,214 @@
-	disney.state = DISNEY_STATE::RUNNING;
-	update_filter_freq(sample_rate);
+	// Keep rendering until we're current
+	while (last_rendered_ms < now) {
+		last_rendered_ms += ms_per_frame;
+		render_queue.emplace(Render());
+	}
 }
 
-// Calculate the sample rate from DAC samples and speed parameters.
-// The maximum possible sample rate is 127,000 Hz which occurs when all 128
-// samples are used and the speedcheck_sum has accumulated only one single
-// tick.
-static uint32_t calc_sample_rate(const dac_channel &dac)
+void Disney::WriteData(io_port_t, const io_val_t data, io_width_t)
 {
-	if (dac.used <= 1)
-		return 0;
-	const uint32_t k_samples = 1000 * (dac.used - 1);
-	const auto sample_rate   = k_samples / dac.speedcheck_sum;
-	return static_cast<uint32_t>(sample_rate);
+	RenderUpToNow();
+	if (!IsFull())
+		fifo.emplace(check_cast<uint8_t>(data));
 }
 
-static void DISNEY_analyze(Bitu channel){
-	switch (disney.state) {
-	case DISNEY_STATE::RUNNING: // should not get here
-		break;
-	case DISNEY_STATE::IDLE:
-		// initialize channel data
-		for (int i = 0; i < 2; i++) {
-			disney.da[i].used = 0;
-			disney.da[i].speedcheck_sum = 0;
-			disney.da[i].speedcheck_failed = false;
-			disney.da[i].speedcheck_init = false;
-		}
-		disney.da[channel].speedcheck_last = PIC_FullIndex();
-		disney.da[channel].speedcheck_init = true;
-
-		disney.state = DISNEY_STATE::ANALYZING;
-		break;
-
-	case DISNEY_STATE::FINISHED: {
-		// The leading channel has the most populated samples
-		disney.leader = disney.da[0].used > disney.da[1].used
-		                        ? &disney.da[0]
-		                        : &disney.da[1];
-
-		// Stereo-mode if both DACs are similarly filled
-		const auto st_diff = abs(disney.da[0].used - disney.da[1].used);
-		disney.stereo = (st_diff < 5);
-
-		// Run with the greater DAC sample rate
-		const auto max_sample_rate = std::max(calc_sample_rate(disney.da[0]),
-		                                      calc_sample_rate(disney.da[1]));
-		DISNEY_enable(max_sample_rate);
-	} break;
-
-	case DISNEY_STATE::ANALYZING: {
-		const auto current = PIC_FullIndex();
-		dac_channel *cch = &disney.da[channel];
-
-		if (!cch->speedcheck_init) {
-			cch->speedcheck_init = true;
-			cch->speedcheck_last = current;
-			break;
-		}
-		const auto speed_delta = current - cch->speedcheck_last;
-		cch->speedcheck_sum += speed_delta;
-		// LOG_MSG("t=%f",current - cch->speedcheck_last);
-
-		// sanity checks (printer...)
-		if (speed_delta < 0.01 || speed_delta > 2)
-			cch->speedcheck_failed = true;
-
-		// if both are failed we are back at start
-		if (disney.da[0].speedcheck_failed && disney.da[1].speedcheck_failed) {
-			disney.state = DISNEY_STATE::IDLE;
-			break;
-		}
-
-		cch->speedcheck_last = current;
-
-		// analyze finish condition
-		if (disney.da[0].used > 30 || disney.da[1].used > 30)
-			disney.state = DISNEY_STATE::FINISHED;
-	} break;
-	}
-}
-
-static void disney_write(io_port_t port, io_val_t value, io_width_t)
```

Comment:
Good riddance! 🔥 Conflating Disney and Covox into the same module, calling it Disney, _then_ doing this runtime analyze stuff to actually revert to Covox behaviour depending on some heuristics was an extremely dodgy and misguided design decision from the get go.

Reply:
Yeah - I started out adding the `AudioFrame RenderFrame()` pattern using the `std::queue<AudioFrame>` approach in the other audio devices.  

But replacing the code's existing FIFO buffer proved quite tricky (even when I included the original code's last-sample click/pop checks).  I was getting some artifacts that weren't in the original code, and wading through the runtime logic (and on-the-fly frequency changes, etc) was getting to be quite involved. 

With that approach not going well, I opted to start from first principles and figured I can add Covox / Generic LPT Stereo DAC as a separate device.
(I figured it's easier to understand this way versus juggling both in the same code)

Reply:
> With that approach not going well, I opted to start from first principles and figured I can add Covox / Generic LPT Stereo DAC as a separate device.

Conflating things and not keeping what's conceptually separate also separate in code is in the "Top 5 Coder Sins of All Time", IMO 😁